File: testdata/languages/python/complex.py
"""
Complex module - tests complexity analysis patterns.
Contains functions with various cyclomatic complexity levels.
"""

from typing import Any, Dict, List, Optional


def simple_function(x: int) -> int:
    """Simple function - complexity 1, no branches."""
    return x * 2


def linear_function(a: int, b: int, c: int) -> int:
    """Linear function - complexity 1, sequential statements."""
    total = a + b
    total += c
    result = total * 2
    return result


def single_if_function(x: int) -> str:
    """Single if - complexity 2."""
    if x > 0:
        return "positive"
    return "non-positive"


def if_else_function(x: int) -> str:
    """If/else - complexity 2."""
    if x > 0:
        return "positive"
    else:
        return "non-positive"


def if_elif_else_function(x: int) -> str:
    """If/elif/else - complexity 3."""
    if x > 0:
        return "positive"
    elif x < 0:
        return "negative"
    else:
        return "zero"


def multiple_if_function(a: bool, b: bool, c: bool) -> int:
    """Multiple independent ifs - complexity 4."""
    count = 0
    if a:
        count += 1
    if b:
        count += 1
    if c:
        count += 1
    return count


def nested_if_function(x: int, y: int) -> str:
    """Nested ifs - complexity 3, nesting 2."""
    if x > 0:
        if y > 0:
            return "both positive"
        return "only x positive"
    return "x not positive"


def deeply_nested_function(a: bool, b: bool, c: bool, d: bool) -> str:
    """Deeply nested - complexity 5, nesting 4."""
    if a:
        if b:
            if c:
                if d:
                    return "all true"
                return "d false"
            return "c false"
        return "b false"
    return "a false"


def match_function(day: int) -> str:
    """Match statement (Python 3.10+) - complexity varies by cases."""
    match day:
        case 1:
            return "Monday"
        case 2:
            return "Tuesday"
        case 3:
            return "Wednesday"
        case 4 | 5:
            return "Thursday or Friday"
        case _:
            return "Weekend"


def for_loop_function(n: int) -> int:
    """For loop - complexity 2, nesting 1."""
    total = 0
    for i in range(n):
        total += i
    return total


def for_each_function(items: List[int]) -> int:
    """For-each loop - complexity 2, nesting 1."""
    total = 0
    for item in items:
        total += item
    return total


def while_loop_function(n: int) -> int:
    """While loop - complexity 2."""
    total = 0
    i = 0
    while i < n:
        total += i
        i += 1
    return total


def nested_loop_function(n: int, m: int) -> int:
    """Nested loops - complexity 3, nesting 2."""
    total = 0
    for i in range(n):
        for j in range(m):
            total += i * j
    return total


def loop_with_condition(items: List[int]) -> int:
    """Loop with condition - complexity 3, nesting 2."""
    total = 0
    for item in items:
        if item > 0:
            total += item
    return total


def loop_with_break(items: List[int], target: int) -> int:
    """Loop with break - complexity 3."""
    for i, item in enumerate(items):
        if item == target:
            return i
    return -1


def loop_with_continue(items: List[int]) -> int:
    """Loop with continue - complexity 3."""
    total = 0
    for item in items:
        if item < 0:
            continue
        total += item
    return total


def logical_and_function(a: bool, b: bool) -> bool:
    """Logical AND - complexity 2 (short-circuit)."""
    return a and b


def logical_or_function(a: bool, b: bool) -> bool:
    """Logical OR - complexity 2 (short-circuit)."""
    return a or b


def complex_logical_function(a: bool, b: bool, c: bool) -> bool:
    """Complex logical - complexity 4."""
    return (a and b) or (b and c) or (a and c)


def ternary_function(x: int) -> str:
    """Ternary operator - complexity 2."""
    return "positive" if x > 0 else "non-positive"


def try_except_function(x: int) -> int:
    """Try/except - complexity 2."""
    try:
        return 100 // x
    except ZeroDivisionError:
        return 0


def try_multiple_except(x: Any) -> int:
    """Try with multiple except - complexity 4."""
    try:
        return int(x)
    except ValueError:
        return 0
    except TypeError:
        return -1
    except Exception:
        return -2


def try_except_finally(x: int) -> int:
    """Try/except/finally - complexity 2."""
    result = 0
    try:
        result = 100 // x
    except ZeroDivisionError:
        result = 0
    finally:
        print("cleanup")
    return result


def with_statement_function(filename: str) -> str:
    """With statement - complexity 1 (no branching)."""
    with open(filename) as f:
        return f.read()


def comprehension_with_condition(items: List[int]) -> List[int]:
    """List comprehension with condition - complexity 2."""
    return [x for x in items if x > 0]


def nested_comprehension(matrix: List[List[int]]) -> List[int]:
    """Nested comprehension - complexity 3."""
    return [x for row in matrix for x in row if x > 0]


def generator_expression_function(items: List[int]) -> int:
    """Generator expression - complexity 2."""
    return sum(x for x in items if x > 0)


def high_complexity_function(
    op: str,
    a: int,
    b: int
) -> Optional[int]:
    """High complexity function - complexity >= 10."""
    if op == "add":
        return a + b
    elif op == "sub":
        return a - b
    elif op == "mul":
        return a * b
    elif op == "div":
        if b == 0:
            return None
        return a // b
    elif op == "mod":
        if b == 0:
            return None
        return a % b
    elif op == "pow":
        if b < 0:
            return None
        result = 1
        for _ in range(b):
            result *= a
        return result
    elif op == "max":
        return a if a > b else b
    elif op == "min":
        return a if a < b else b
    else:
        return None


def extremely_complex_function(
    category: str,
    items: List[Dict[str, Any]],
    threshold: int
) -> List[Dict[str, Any]]:
    """Extremely complex function - complexity >= 15, nesting >= 4."""
    if not items:
        raise ValueError("empty items")

    result = []
    for item in items:
        if item.get("category") != category:
            continue

        status = item.get("status")
        if status == "active":
            if item.get("value", 0) > threshold:
                if item.get("priority") == "high":
                    result.append(item)
                elif item.get("priority") == "medium":
                    if item.get("value", 0) > threshold * 2:
                        result.append(item)
        elif status == "pending":
            if item.get("created_at"):
                for tag in item.get("tags", []):
                    if tag == "important":
                        result.append(item)
                        break
        elif status == "archived":
            continue
        else:
            if item.get("value", 0) > 0 and item.get("priority"):
                result.append(item)

    if not result:
        raise ValueError(f"no items found for category {category}")

    return result


async def async_complex_function(
    urls: List[str],
    timeout: int
) -> List[str]:
    """Async complex function - complexity >= 5."""
    results = []
    for url in urls:
        if not url:
            continue
        try:
            # Simulated async operation
            if timeout > 0:
                results.append(f"fetched: {url}")
            else:
                results.append(f"timeout: {url}")
        except Exception as e:
            results.append(f"error: {e}")
    return results


class ComplexClass:
    """Class with complex methods for testing."""

    def __init__(self, config: Dict[str, Any]):
        self.config = config
        self._initialized = False

    def complex_method(
        self,
        items: List[int],
        mode: str
    ) -> int:
        """Complex method - high complexity."""
        if not self._initialized:
            return -1

        total = 0
        for item in items:
            if mode == "sum":
                total += item
            elif mode == "product":
                if total == 0:
                    total = 1
                total *= item
            elif mode == "max":
                if item > total:
                    total = item
            elif mode == "min":
                if total == 0 or item < total:
                    total = item
            else:
                continue

        return total

File: testdata/languages/go/complex.go
// Package main - complex.go tests complexity analysis patterns.
// Contains functions with various cyclomatic complexity levels.
package main

import (
	"errors"
	"fmt"
)

// SimpleFunction has complexity 1 - no branches.
// Expected: cyclomatic complexity = 1, nesting = 0
func SimpleFunction(x int) int {
	return x * 2
}

// LinearFunction has complexity 1 - sequential statements.
// Expected: cyclomatic complexity = 1, nesting = 0
func LinearFunction(a, b, c int) int {
	sum := a + b
	sum += c
	result := sum * 2
	return result
}

// SingleIfFunction has complexity 2 - one if statement.
// Expected: cyclomatic complexity = 2, nesting = 1
func SingleIfFunction(x int) string {
	if x > 0 {
		return "positive"
	}
	return "non-positive"
}

// IfElseFunction has complexity 2 - if/else.
// Expected: cyclomatic complexity = 2, nesting = 1
func IfElseFunction(x int) string {
	if x > 0 {
		return "positive"
	} else {
		return "non-positive"
	}
}

// MultipleIfFunction has complexity 4 - multiple independent ifs.
// Expected: cyclomatic complexity = 4, nesting = 1
func MultipleIfFunction(a, b, c bool) int {
	count := 0
	if a {
		count++
	}
	if b {
		count++
	}
	if c {
		count++
	}
	return count
}

// NestedIfFunction has complexity 3 - nested ifs.
// Expected: cyclomatic complexity = 3, nesting = 2
func NestedIfFunction(x, y int) string {
	if x > 0 {
		if y > 0 {
			return "both positive"
		}
		return "only x positive"
	}
	return "x not positive"
}

// DeeplyNestedFunction has high nesting depth.
// Expected: cyclomatic complexity = 5, nesting = 4
func DeeplyNestedFunction(a, b, c, d bool) string {
	if a {
		if b {
			if c {
				if d {
					return "all true"
				}
				return "d false"
			}
			return "c false"
		}
		return "b false"
	}
	return "a false"
}

// SwitchFunction has complexity based on cases.
// Expected: cyclomatic complexity = 5 (4 cases + default)
func SwitchFunction(day int) string {
	switch day {
	case 1:
		return "Monday"
	case 2:
		return "Tuesday"
	case 3:
		return "Wednesday"
	case 4, 5:
		return "Thursday or Friday"
	default:
		return "Weekend"
	}
}

// ForLoopFunction has complexity 2 - one loop.
// Expected: cyclomatic complexity = 2, nesting = 1
func ForLoopFunction(n int) int {
	sum := 0
	for i := 0; i < n; i++ {
		sum += i
	}
	return sum
}

// ForRangeFunction has complexity 2 - range loop.
// Expected: cyclomatic complexity = 2, nesting = 1
func ForRangeFunction(items []int) int {
	sum := 0
	for _, item := range items {
		sum += item
	}
	return sum
}

// NestedLoopFunction has complexity 3 - nested loops.
// Expected: cyclomatic complexity = 3, nesting = 2
func NestedLoopFunction(n, m int) int {
	sum := 0
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			sum += i * j
		}
	}
	return sum
}

// LoopWithCondition has complexity 3 - loop + condition.
// Expected: cyclomatic complexity = 3, nesting = 2
func LoopWithCondition(items []int) int {
	sum := 0
	for _, item := range items {
		if item > 0 {
			sum += item
		}
	}
	return sum
}

// LogicalAndFunction has complexity 2 - && operator.
// Expected: cyclomatic complexity = 2 (short-circuit)
func LogicalAndFunction(a, b bool) bool {
	return a && b
}

// LogicalOrFunction has complexity 2 - || operator.
// Expected: cyclomatic complexity = 2 (short-circuit)
func LogicalOrFunction(a, b bool) bool {
	return a || b
}

// ComplexLogicalFunction has complexity 4 - multiple logical ops.
// Expected: cyclomatic complexity = 4
func ComplexLogicalFunction(a, b, c bool) bool {
	return (a && b) || (b && c) || (a && c)
}

// ErrorHandlingFunction has complexity 4 - multiple error checks.
// Expected: cyclomatic complexity = 4, nesting = 1
func ErrorHandlingFunction(x, y int) (int, error) {
	if x < 0 {
		return 0, errors.New("x must be non-negative")
	}
	if y < 0 {
		return 0, errors.New("y must be non-negative")
	}
	if y == 0 {
		return 0, errors.New("y must not be zero")
	}
	return x / y, nil
}

// HighComplexityFunction has high cyclomatic complexity.
// Expected: cyclomatic complexity >= 10
func HighComplexityFunction(op string, a, b int) (int, error) {
	switch op {
	case "add":
		return a + b, nil
	case "sub":
		return a - b, nil
	case "mul":
		return a * b, nil
	case "div":
		if b == 0 {
			return 0, errors.New("division by zero")
		}
		return a / b, nil
	case "mod":
		if b == 0 {
			return 0, errors.New("modulo by zero")
		}
		return a % b, nil
	case "pow":
		if b < 0 {
			return 0, errors.New("negative exponent")
		}
		result := 1
		for i := 0; i < b; i++ {
			result *= a
		}
		return result, nil
	case "max":
		if a > b {
			return a, nil
		}
		return b, nil
	case "min":
		if a < b {
			return a, nil
		}
		return b, nil
	default:
		return 0, fmt.Errorf("unknown operation: %s", op)
	}
}

// TypeSwitchFunction has complexity based on type cases.
// Expected: cyclomatic complexity = 4
func TypeSwitchFunction(v interface{}) string {
	switch v := v.(type) {
	case int:
		return fmt.Sprintf("int: %d", v)
	case string:
		return fmt.Sprintf("string: %s", v)
	case bool:
		return fmt.Sprintf("bool: %t", v)
	default:
		return "unknown type"
	}
}

// DeferFunction tests defer statement.
// Expected: cyclomatic complexity = 2
func DeferFunction(shouldPanic bool) (result string) {
	defer func() {
		if r := recover(); r != nil {
			result = "recovered"
		}
	}()

	if shouldPanic {
		panic("intentional panic")
	}

	return "normal"
}

// GoroutineFunction tests goroutine spawning.
// Expected: cyclomatic complexity = 2
func GoroutineFunction(items []int) <-chan int {
	ch := make(chan int)
	go func() {
		defer close(ch)
		for _, item := range items {
			ch <- item
		}
	}()
	return ch
}

// SelectFunction tests select statement.
// Expected: cyclomatic complexity = 4
func SelectFunction(ch1, ch2 <-chan int, done <-chan struct{}) int {
	select {
	case v := <-ch1:
		return v
	case v := <-ch2:
		return v * 2
	case <-done:
		return -1
	default:
		return 0
	}
}

// ExtremelyComplexFunction is intentionally complex for testing.
// Expected: cyclomatic complexity >= 15, nesting >= 4
func ExtremelyComplexFunction(category string, items []Item, threshold int) ([]Item, error) {
	if len(items) == 0 {
		return nil, errors.New("empty items")
	}

	var result []Item
	for _, item := range items {
		if item.Category != category {
			continue
		}

		switch item.Status {
		case "active":
			if item.Value > threshold {
				if item.Priority == "high" {
					result = append(result, item)
				} else if item.Priority == "medium" {
					if item.Value > threshold*2 {
						result = append(result, item)
					}
				}
			}
		case "pending":
			if item.CreatedAt > 0 {
				for _, tag := range item.Tags {
					if tag == "important" {
						result = append(result, item)
						break
					}
				}
			}
		case "archived":
			// Skip archived items
			continue
		default:
			if item.Value > 0 && item.Priority != "" {
				result = append(result, item)
			}
		}
	}

	if len(result) == 0 {
		return nil, fmt.Errorf("no items found for category %s", category)
	}

	return result, nil
}

// Item represents a data item for testing.
type Item struct {
	Category  string
	Status    string
	Value     int
	Priority  string
	CreatedAt int64
	Tags      []string
}
